### PR TITLE
feat(transactions): Add app_start_type migration

### PR DESCRIPTION
**Pull request by philipphofmann**
Adds the column app_start_type to transactions. For more context,
see https://github.com/getsentry/sentry/issues/36927.

### CODE REVIEW

**Comment by philipphofmann** [comment]
@evanh, are there any further actions required to make that searchable in Discover?

File: snuba/datasets/transactions_processor.py [hunk 1]
```diff
@@ -248,6 +248,8 @@ def _process_contexts_and_user(
             if context in contexts:
                 del contexts[context]
 
+        processed["app_start_type"] = contexts["app"]["start_type"]
```
Comment (evanh) [inline snuba/datasets/transactions_processor.py:251]: Is this field guaranteed to always exist?

File: snuba/datasets/entities/discover.py [hunk 2]
```diff
@@ -284,6 +284,7 @@ def attempt_map(
                 ]
             ),
         ),
+        ("app_start_type", String(Modifiers(nullable=True))),
```
Comment (evanh) [inline snuba/datasets/entities/discover.py:287]: I would add/modify an E2E test in `test_snql_api.py` or `test_snql_sdk_api.py` that ensures this column can be read correctly.

**Review by evanh — COMMENTED** [review]
Please separate the migration change from the other changes. Currently migrations in Snuba are run manually, so if this is merged in the current state inserts will start failing because the column doesn't exist on the table. The steps for adding a new column are:

1) Open a migration PR and a second PR for the reading/writing changes.
2) Merge the migration PR
3) Someone from SnS will manually run the migration
4) Merge the second PR.

> are any further actions required to make that searchable in Discover?

The column also needs to be exposed in Sentry, via https://github.com/getsentry/sentry/blob/master/src/sentry/snuba/events.py

File: snuba/datasets/transactions_processor.py (hunk 1 above)
Comment (philipphofmann) [reply snuba/datasets/transactions_processor.py:251]: No, definitely not.

**Comment by philipphofmann** [comment]
Thanks for the feedback, @evanh. I put everything together in one PR to be able to get feedback at one glance. I'm going to split the PR now.

File: snuba/datasets/transactions_processor.py (hunk 1 above)
Comment (philipphofmann) [reply snuba/datasets/transactions_processor.py]: Ah, I just learned the difference between `dict[key]` and `dict.get(key)`. I'm new to Python.

File: snuba/datasets/storages/transactions_common.py [hunk 3]
```diff
@@ -101,6 +101,7 @@
         ("transaction_source", String()),
         ("timestamp", DateTime(Modifiers(readonly=True))),
         ("group_ids", Array(UInt(64))),
+        ("app_start_type", String(Modifiers(readonly=True))),
```
Comment (nikhars) [inline snuba/datasets/storages/transactions_common.py:104]: Why do you add `readonly=True` modifier to this column? Is this column going to be materialized? If not, I would recommend that you remove `readonly=True` modifier

File: snuba/snuba_migrations/transactions/0017_transactions_add_app_start_type_column.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    def __forward_migrations(
+        self, table_name: str
+    ) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.AddColumn(
+                storage_set=StorageSetKey.TRANSACTIONS,
+                table_name=table_name,
+                column=Column(
+                    "app_start_type",
+                    String(Modifiers(low_cardinality=True)),
```
Comment (nikhars) [inline snuba/snuba_migrations/transactions/0017_transactions_add_app_start_type_column.py:26]: Is app_start_type field user generated or machine generated? If it is machine generated and would have mostly a few values then having it as low_cardinality is ok. If it is user generated, what is the likelihood of having high cardinality of this value?

Comment (philipphofmann) [reply snuba/snuba_migrations/transactions/0017_transactions_add_app_start_type_column.py:26]: The SDKs create these values, and it's not exposed to the user. Currently, we have four different values. The likelihood of this having a high cardinality is very low. Maybe it increases to 10 or 20 at some point, but not more than 100 for sure.

File: snuba/datasets/storages/transactions_common.py (hunk 3 above)
Comment (philipphofmann) [reply snuba/datasets/storages/transactions_common.py:104]: What exactly do you mean by `materialized`? I am OK with removing it.

**Pull request merged**